Add Floating Menu

usulpro opened the pull request:
It implements #56

I moved all items from 'keyboard shortcuts' to menu wich avalibale in the left panel. And when the left panel is hidden it floats in the preview area. It also has buttons to browse through stories. It gives access to Storybook core functionalities both by a mouse and a touchscreen from normal and fullscreen view.

here is a [demo page](https://usulpro.github.io/storybook-ui) to check it on mobile device.

usulpro committed:
windows support

usulpro committed:
Floating menu

usulpro committed:
menu position

usulpro committed:
add jumpToKind

usulpro committed:
enlarge buttons for touchscreens

arunoda commented:
This is great. Here's a list of common UI suggestions.

* Add a way to show our existing keymap window. (Add a new section called Keyboard shortcuts in the menu)
* I wanna see this on the top in the normal view: http://recordit.co/499hST1YVy

arunoda commented on src/modules/ui/components/floating_menu/svg_package.js:
```diff
@@ -0,0 +1,165 @@
+/* eslint-disable */
```
Can't we get this via NPM?

UsulPro replied:
It's just a bundle with svg images which I use in my menu. I have two reasons to put them into one file:

- to don't bloat src with a bunch of image files

- to don't add new dependency to project (I need here `file-loader` or `base64-image-loader`)

So I just prebuild it manually with `base64-image-loader`. Maybe it could be usefull later if you'll want to extend it with more images? But if you don't like such approach could you suggest the better way how to get this images?

arunoda commented on src/modules/ui/routes.js:
```diff
@@ -15,12 +16,15 @@ export default function (injectDeps, { clientStore, provider, domNode }) {
     <div>
       <Layout
-        leftPanel={() => (<LeftPanel />)}
+        leftPanel={(menu) => (<LeftPanel floatingMenu={menu} />)}
```
I don't like getting the menu from a function. Since we define the floatingMenu inside here. Let's use it directly.

UsulPro replied:
Ok I removed it from here

arunoda commented:
@mnmtanish could you review this too?

mnmtanish commented:
Sure, I'll check it out today evening.

usulpro committed:
update

UsulPro commented:
I fixed it according your suggestion and updated [demo page](https://usulpro.github.io/storybook-ui/).

- I added new menu item "Keyboard shortcuts". But the keymap window is visible only in normal view. How do you think is it Ok? (BTW did you noticed that if you hold mouse over any menu item, the tooltip with keyboard shortcut will be shown?) Another idea to show keyboard shortcuts may be to change menu item names to shortcuts when the mouse is over this menu item.

- I moved it as you asked and make menu appearing from below (I'll rename 'Footer' to 'Menu' then)

- I sped up transition to 100ms but if you still don't like it I can remove it complitely

mnmtanish commented on src/modules/api/actions/api.js:
```diff
@@ -24,6 +24,21 @@ export function jumpToStory(storyKinds, selectedKind, selectedStory, direction)
 
+function jumpToKind(storyKinds, selectedKind, selectedStory, direction) {
+  const currentIndex = storyKinds.findIndex(({ kind }) => (kind === selectedKind));
+  if (currentIndex === -1) return selectedKind;
```
Shouldn't we return both `selectedKind` and `selectedStory` ?

UsulPro replied:
sure! fixed :)

mnmtanish commented on src/modules/ui/components/left_panel/header.js:
```diff
@@ -42,14 +43,15 @@ const shortcutIconStyle = {
 };
 
-const Header = ({ openShortcutsHelp, name, url }) => (
+const Header = ({/* openShortcutsHelp,*/ name, url }) => (
```
Just delete unused code, we can always `git checkout` if we need them again

mnmtanish replied:
Found some more below, check them too

UsulPro replied:
Ok

mnmtanish commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
 node_modules
+dist
```
Move this to a new PR, this one needs to be discussed more. Having `dist` on the repo sucks but we did it anyway so that storybook could be installed from the Github repo

UsulPro replied:
Oh, no problem!
I just revert it and make `npm run prepublish` before pushing

usulpro committed:
update2

usulpro committed:
update2.tests

UsulPro commented on dist/libs/key_events.js:
```diff
@@ -14,15 +14,17 @@ var _keycode2 = _interopRequireDefault(_keycode);
-  PREV_STORY: 7,
-  SEARCH: 8,
-  DOWN_PANEL_IN_RIGHT: 9
+  FULLSCREEN: 'FULLSCREEN',
```
I changed it to strings because it was hard to setup tests for that. Now all test passed so if it matters I can revert it if you want!?

UsulPro replied:
I changed it to strings because it was hard to setup tests for that. Now all test passed so if it matters I can revert it if you want!?

UsulPro commented on src/modules/ui/components/floating_menu/index.js:
```diff
@@ -0,0 +1 @@
+export FloatingMenu from './floating_menu';
```
ESLint complains here. Should I disable it here?

UsulPro commented:
BTW don't you want to move 'text filter' also into this menu? (or add an item to hide it)
I can add this if you like it

UsulPro commented:
@arunoda, @mnmtanish!
Is it possible to merge this?

ndelangen commented:
Let's create a PR here: https://github.com/storybooks/storybook
